### PR TITLE
[B] Make player argument optional

### PR DESCRIPTION
**Pull request by stuntguy3000**
**The Issue:**
I believe for /playSound the player parameter should be optional, and default to the sender.

**Justification for this PR:**
This PR was made to help improve command usage (ease of use), to remove unnecessary code and mainly to improve command speed/efficiency. 

For example, If i wanted to test a certain sound (to see if it matches my needs or not), (I have a list of all of the sound names) I would get the sound name, then paste it in in-game into the command syntax. This is good and dandy, until you start demoing a lot of sounds. 

I would really have two options when sampling a lot of sounds
1) Have to back arrow and remove the sound ID in chat yourself
2) Modify the list to have the command syntax in place (As in, put **/playsound** in front and **stuntguy3000** behind the sound ID.

I currently do the 1st option, as the 2nd option is rather slow, tedious and It really messes up the look of the document. 

**In summary: This PR is designed to make the command easier and simpler to use, as well as increasing speed and efficiency.**

**Description of changes:**
Each change is commented. Scroll down to view.

**JIRA Ticket:**
https://bukkit.atlassian.net/browse/BUKKIT-5374

**Testing Material**
Compiled changes from local fork and tested each argument.

Associated Bukkit PR: https://github.com/Bukkit/Bukkit/pull/1026

### CODE REVIEW

File: src/main/java/org/bukkit/command/defaults/PlaySoundCommand.java [hunk 1]
```diff
@@ -10,7 +10,7 @@
     public PlaySoundCommand() {
         super("playsound");
         this.description = "Plays a sound to a given player";
-        this.usageMessage = "/playsound <sound> <player> [x] [y] [z] [volume] [pitch] [minimumVolume]";
+        this.usageMessage = "/playsound <sound> [player] [x] [y] [z] [volume] [pitch] [minimumVolume]";
```
Comment (stuntguy3000) [inline src/main/java/org/bukkit/command/defaults/PlaySoundCommand.java:13]: Specify this parameter is now optional.

File: src/main/java/org/bukkit/command/defaults/PlaySoundCommand.java [hunk 2]
```diff
@@ -20,12 +20,19 @@ public boolean execute(CommandSender sender, String currentAlias, String[] args)
             return true;
         }
 
-        if (args.length < 2) {
+        if (args.length < 1) {
```
Comment (stuntguy3000) [inline src/main/java/org/bukkit/command/defaults/PlaySoundCommand.java]: Change from 2 required arguments to 1.

File: src/main/java/org/bukkit/command/defaults/PlaySoundCommand.java [hunk 3]
```diff
@@ -20,12 +20,19 @@ public boolean execute(CommandSender sender, String currentAlias, String[] args)
...
             sender.sendMessage(ChatColor.RED + "Usage: " + usageMessage);
             return false;
         }
+
```
Comment (stuntguy3000) [inline src/main/java/org/bukkit/command/defaults/PlaySoundCommand.java:34]: Formatting fix.

File: src/main/java/org/bukkit/command/defaults/PlaySoundCommand.java [hunk 4]
```diff
@@ -20,12 +20,19 @@ public boolean execute(CommandSender sender, String currentAlias, String[] args)
...
         final String soundArg = args[0];
-        final String playerArg = args[1];
+        String playerArg = null;
+
+        if (args.length == 1) {
+            playerArg = sender.getName();
+        } else {
+            playerArg = args[1];
+        }
```
Comment (stuntguy3000) [inline src/main/java/org/bukkit/command/defaults/PlaySoundCommand.java:42]: If 1 argument is supplied, Default to sender's username, if not function like before (by using entered username).

File: src/main/java/org/bukkit/command/defaults/PlaySoundCommand.java [hunk 5]
```diff
@@ -56,8 +62,6 @@ public boolean execute(CommandSender sender, String currentAlias, String[] args)
             y = getRelativeDouble(y, sender, args[3]);
         case 3:
             x = getRelativeDouble(x, sender, args[2]);
-        case 2:
-            // Noop
```
Comment (stuntguy3000) [inline src/main/java/org/bukkit/command/defaults/PlaySoundCommand.java:60]: For changes above, Removed unnecessary code.

**Comment by turt2live** [comment]
What is the issue with making a plugin that does this sole purpose?

Modifying the command arguments is not really appropriate for the issue you are presenting.

**Comment by gravitylow** [comment]
This doesn't make sense for usage by the console.

**Comment by stuntguy3000** [comment]
**This doesn't make sense for usage by the console.**
What do you mean? You would specify the player..would you not?

**What is the issue with making a plugin that does this sole purpose?**

**Modifying the command arguments is not really appropriate for the issue you are presenting.**
I did present my justification poorly, The plugin i'm making use's a whole range of sounds, however I don't know the right sound names/pitches. I think this minor change is helpful so..

So I can simply run "/effect ", let it fail, then simply press the Up key, Paste in the name, and see how it sounds (Faster command execution). Speed is key.

**Comment by turt2live** [comment]
Again, convenience is not a decent justification for a PR like this.

Additionally, you have not investigated nor fixed the issue presented by @gravitylow that you introduced with this PR.

**Comment by turt2live** [comment]
I'd also like to point out that it is entirely possible to create a new text file to contain the commands temporarily and copy/paste the commands (with new lines) into the console.

`playsound whatever turt2live ... etc
playsound whatever turt2live ... etc
playsound whatever turt2live ... etc
playsound whatever turt2live ... etc
playsound whatever turt2live ... etc`

Copy/paste into console and magic happens.

**Comment by stuntguy3000** [comment]
"Again, convenience is not a decent justification for a PR like this."

Why isn't it?

**Comment by turt2live** [comment]
Because you haven't given a reason besides "I need this".

sent from mobile

**Comment by stuntguy3000** [comment]
I have, actually.

My key point is really this change allows **efficient and faster** usage of the command.

**Comment by turt2live** [comment]
And I've brought up concerns, plenty of them.

Feel free to enhance the justification and repair the error in
communication introduced by this PR.

sent from mobile

**Comment by stuntguy3000** [comment]
Updated Justification, and fixed issues. Waiting for feedback.

**Comment by ryantheleach** [comment]
Stuntguy, if vanilla doesn't do this, neither should bukkit. (on the flip side if vanilla DOES have this behavior, say so in your pull request.

**Comment by stuntguy3000** [comment]
There is **a lot of things that Bukkit has (Vanilla command remakes)** that I can apply that argument too. I don't see how it is valid in the slightest.

**Comment by stuntguy3000** [comment]
Didn't mean to close,

File: src/main/java/org/bukkit/command/defaults/PlaySoundCommand.java [hunk 6]
```diff
@@ -20,12 +20,26 @@ public boolean execute(CommandSender sender, String currentAlias, String[] args)
...
-            sender.sendMessage(ChatColor.RED + "Usage: " + usageMessage);
-            return false;
+        if (sender instanceof Player) {
+            if (args.length < 1) {
+                sender.sendMessage(ChatColor.RED + "Usage: " + usageMessage);
+                return false;
+            }
+        } else {
+            if (args.length < 2) {
+                sender.sendMessage(ChatColor.RED + "Usage: /playsound <sound> <player> [x] [y] [z] [volume] [pitch] [minimumVolume]");
```
Comment (turt2live) [inline src/main/java/org/bukkit/command/defaults/PlaySoundCommand.java:30]: This is incorrect usage of a usage message.This is what can potentially cause a mistranslated command in the future.

Comment (stuntguy3000) [inline src/main/java/org/bukkit/command/defaults/PlaySoundCommand.java:30]: How is this incorrect?

**Comment by gravitylow** [comment]
@stuntguy3000 my comment means by defaulting the player to the command sender, you would, by default, be playing the sound to a console if the command was entered there. That does not make sense.

File: src/main/java/org/bukkit/command/defaults/PlaySoundCommand.java (hunk 6 above)
Comment (turt2live) [inline src/main/java/org/bukkit/command/defaults/PlaySoundCommand.java:30]: You assume it will never change

Comment (stuntguy3000) [inline src/main/java/org/bukkit/command/defaults/PlaySoundCommand.java:30]: I don't see the issue you can see, How would you fix what you are describing?

**Comment by stuntguy3000** [comment]
@gravitylow I fixed the issue you mentioned, by adding in a check.

**Comment by EvilSeph** [comment]
We aim to provide Minecraft vanilla's command behaviour at the VERY LEAST, with optional improvements on top of that. Changing the core behaviour of Minecraft's commands falls outside of the scope of this project.

**Comment by stuntguy3000** [comment]
How is this changing the core behavior? This is a very small improvement, is it not?
